pkg/authorization/admission/restrictusers/restrictusers.go

```diff
@@ -0,0 +1,353 @@
+package restrictusers
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/admission"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/labels"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+	oadmission "github.com/openshift/origin/pkg/cmd/server/admission"
+	"github.com/openshift/origin/pkg/controller/shared"
+)
+
+const (
+	// CollaboratorAnnotation is the key for an annotation on Namespace that
+	// specifies collaborators on the namespace.
+	CollaboratorAnnotation = "authorization.openshift.io/collaborators"
```

Comment:
In 3.3, the kube infrastructure to register different groups was hopelessly broken.  In 3.4, it's only mostly broken.  In 3.5, it mostly works.

I think I'd prefer to see this as a separate API group and resource.  That would allow "clean" management via `oc` and all of our usual tooling.